### PR TITLE
Support `git` pubspec source, cleanup 

### PR DESCRIPTION
**Pull request by ckiee**
hi hi~!
###### TODO
- [x] deps2nix: `git` support
- [x] build-{flutter,dart}-app.nix: parse new deps2nix lock format
- [x] deps2nix: rename to flutter2nix (pubspec2nix?)

### CODE REVIEW

**Comment by gilice** [comment]
hi! sorry for bothering. any progress on this?

**Comment by ckiee** [comment]
nope, but @FlafyDev's taking over this after your comment. (discussed out-of-band)

File: pubspec-nix/pubspec-nix.py [hunk 1]
```diff
@@ -192,7 +193,7 @@ def main():
 
     pubspec_lock = yaml.safe_load(open("pubspec.lock", "r"))
 
-    pbar = tqdm(total=len({ k: v for k, v in pubspec_lock["packages"].items() if "url" in v["description"] }))
+    pbar = tqdm(total=len({ k: v for k, v in pubspec_lock["packages"].items() if v["source"] == "git" or v["source"] == "hosted" }))
```
Comment (ckiee) [inline pubspec-nix/pubspec-nix.py:196]: it'd be nicer semantically if the program knew about this more generally

File: builders/shared/default.nix [hunk 2]
```diff
@@ -1,17 +1,25 @@
...
-    (path: dep: let
-      derv = fetchzip dep;
-    in ''
-      mkdir -p $out/${dirOf path}
-      ln -s ${derv} $out/${path}
-    '')
-    deps.pub));
+  }: (runCommand "${args.pname}-pub-cache" {} (lib.mapAttrsToList (
+      path: dep: let
+        derv = pkgs.${dep.fetcher} dep.args;
```
Comment (ckiee) [inline builders/shared/default.nix:11]: this is scary to me: i can't think of how exactly it'd be bad but it's untrusted-ish data being fed into nix where there's surely a way to make that bad.

<suggestion supplied out-of-band>

Comment (FlafyDev) [reply builders/shared/default.nix]: done. i used your idea of having a fetchers attrset.

File: pubspec-nix/pubspec-nix.py (hunk 1 above)
Comment (FlafyDev) [reply pubspec-nix/pubspec-nix.py]: done

File: builders/shared/default.nix [hunk 3]
```diff
@@ -1,25 +1,31 @@
...
-        + (lib.optionalString (dep.fetcher == "fetchgit") ''
-          local CACHE_DIR=$out/git/cache/${dep.repo_name}-$(echo -n ${dep.args.url} | sha1sum | awk '{print $1}')
-          mkdir -p $(dirname $CACHE_DIR)
-          ln -s ${derv}/.git $CACHE_DIR
-        '')
-    )
-    pubspecNixLock.pub));
+  generatePubCache = let
+    getFetcher = fetcher:
+      {inherit fetchzip fetchgit;}.${fetcher}
```
Comment (ckiee) [inline builders/shared/default.nix:9]: Some spacing would be nice. Rerun your formatter? (`{ inherit …; }`)

File: builders/shared/default.nix [hunk 4]
```diff
@@ -1,25 +1,31 @@
...
+      or (throw "Unknown fetcher: ${fetcher}");
```
Comment (ckiee) [inline builders/shared/default.nix:10]:
```suggestion
      or throw "Unknown fetcher: ${fetcher}";
 ```

File: builders/shared/default.nix [hunk 5]
```diff
@@ -1,25 +1,31 @@
...
-        # https://github.com/dart-lang/pub/blob/12a2af4de1e85994eae5679c7a7ed082c5a557f5/lib/src/source/git.dart#L338
-        + (lib.optionalString (dep.fetcher == "fetchgit") ''
-          local CACHE_DIR=$out/git/cache/${dep.repo_name}-$(echo -n ${dep.args.url} | sha1sum | awk '{print $1}')
-          mkdir -p $(dirname $CACHE_DIR)
-          ln -s ${derv}/.git $CACHE_DIR
-        '')
-    )
-    pubspecNixLock.pub));
+  generatePubCache = let
+    getFetcher = fetcher:
```
Comment (ckiee) [inline builders/shared/default.nix:8]: It could also use dep.fetcher directly, up to you.

File: builders/shared/default.nix (hunk 3 above)
Comment (FlafyDev) [reply builders/shared/default.nix:9]: I am using alejandra as a formatter, it likes to remove these spacing

File: builders/shared/default.nix (hunk 4 above)
Comment (FlafyDev) [reply builders/shared/default.nix:10]: removing the parentheses leads to a `error: value is a string while a set was expected`

File: builders/shared/default.nix (hunk 5 above)
Comment (FlafyDev) [reply builders/shared/default.nix]: done

File: builders/shared/default.nix (hunk 4 above)
Comment (ckiee) [reply builders/shared/default.nix]: huh!

**Comment by FlafyDev** [comment]
Merged. Thanks!

**Pull request merged**